### PR TITLE
Remove launch from expt + add sycl backend + dynamic shared memory

### PR DESCRIPTION
**Pull request by artv3**
Introduces the sycl backend for launch and dynamic shared memory (for all backends).  This PR will introduce API changes.

- [x] Keep prototype of perfectly nested loop interface in expt.

### CODE REVIEW

File: include/RAJA/policy/sycl/teams.hpp
Comment (artv3) [inline include/RAJA/policy/sycl/teams.hpp:35]: @rchen20 do you have any thoughts on what the value to the right of async should be? It was switched to 1 for cuda, but not sure what it should be here.

Comment (rchen20) [reply include/RAJA/policy/sycl/teams.hpp:35]: I had purposely not implemented a `sycl_launch` policy because I'm not sure what the equivalent of "blocks per SM" would be in SYCL, or if SYCL can support a blocks per SM input upon launch. Perhaps @homerdin would know?

Comment (MrBurmark) [reply include/RAJA/policy/sycl/teams.hpp:35]: What would you do in sycl? The policies express what is available in the backend (sycl) not necessarily what's available in the HW or implementation.

Comment (trws) [reply include/RAJA/policy/sycl/teams.hpp:35]: What do we need that for? SYCL doesn't expose that directly AFAIK, but you can query all the necessary properties to compute it. At least you can through the underlying opencl, these are the two I would look to for this (CL_​DEVICE_​MAX_​COMPUTE_​UNITS probably most relevant, combined with kernel info for how many can actually be resident):

* https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/clGetDeviceInfo.html
* https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/clGetKernelWorkGroupInfo.html

**Comment by artv3** [comment]
Hi all, 

I feel this PR might be in pretty good shape to start reviewing. I do want to propose an additional API change, should we change Teams -> Blocks in the Grid data structure to be more consistent with CUDA/HIP? 

Second, I have a prototype for perfectly nested loops here, its not very flexible though, my thought was to remove what is in there and follow up with a proper implementation. 

ping @MrBurmark , @rchen20 , @tomstitt , @rhornung67

**Comment by artv3** [comment]
To check out the dynamic shared memory I suggest looking at the following example: dynamic_mat_transpose.cpp 
It has support for seq/openmp/cuda/hip/sycl.

**Comment by MrBurmark** [comment]
> Second, I have a prototype for perfectly nested loops here, its not very flexible though, my thought was to remove what is in there and follow up with a proper implementation.

If its just a prototype you could leave it in expr perhaps?

**Comment by artv3** [comment]
> > Second, I have a prototype for perfectly nested loops here, its not very flexible though, my thought was to remove what is in there and follow up with a proper implementation.
> 
> If its just a prototype you could leave it in expr perhaps?

Yes, lets do that.

File: examples/dynamic_mat_transpose.cpp
Comment (MrBurmark) [inline examples/dynamic_mat_transpose.cpp:367]: What does this do? Is the memory not allocated before we enter the body of the launch?

File: examples/launch_flatten.cpp
Comment (MrBurmark) [inline examples/launch_flatten.cpp:106]: Does launch now always require shmem as an arg? It would be nice if it was optional.
Maybe if it used a special type it would be easier to make optional?
```suggestion
    (RAJA::DynamicSharedMemory(dynamic_shared_mem), grid, [=] RAJA_HOST_DEVICE (RAJA::LaunchContext ctx)
```

File: include/RAJA/policy/loop/launch.hpp
Comment (MrBurmark) [inline include/RAJA/policy/loop/launch.hpp:46]: I assume you want uninitialized memory here? I think malloc is a more direct way of getting that.

File: include/RAJA/policy/sycl/launch.hpp
Comment (MrBurmark) [inline include/RAJA/policy/sycl/launch.hpp:77]: Does it make more sense to make this lambda mutable so you don't have to make the members of LaunchContext mutable?

File: examples/launch_flatten.cpp
Comment (artv3) [reply examples/launch_flatten.cpp:106]: yes, I like both suggestions.

Comment (artv3) [reply examples/launch_flatten.cpp:106]: How can we make it optional with the special type? I'm thinking it would be easy to make optional if it was the last argument of launch, but not sure if it was not.

Comment (rhornung67) [reply examples/launch_flatten.cpp:106]: @artv3 would it help or make sense to pass the shmem to the LaunchContext ctor, optionally?

File: examples/dynamic_mat_transpose.cpp
Comment (MrBurmark) [reply examples/dynamic_mat_transpose.cpp:367]: Oh, I see its required because of the the offsetting. Indeed there is no way to auto reset that given how it works.

File: examples/launch_flatten.cpp
Comment (MrBurmark) [reply examples/launch_flatten.cpp:106]: Or maybe have a single general front-end that takes a parameter pack when we could have a function that would find a specific type from the parameter pack or return a default if not found.

Comment (rhornung67) [reply examples/launch_flatten.cpp:106]: That would work too.  To me, it seems that LaunchContext should take on more responsibilities.

File: examples/dynamic_mat_transpose.cpp
Comment (MrBurmark) [inline examples/dynamic_mat_transpose.cpp:334]: Should this just be called Tile like in the c-style example?

File: examples/dynamic_mat_transpose.cpp
Comment (MrBurmark) [inline examples/dynamic_mat_transpose.cpp:367]: Do you want to add a comment here with more explanation of why this is necessary?

Comment (artv3) [reply examples/dynamic_mat_transpose.cpp:367]: yes, will do.

File: examples/launch_flatten.cpp
Comment (artv3) [reply examples/launch_flatten.cpp:106]: Actually passing it into the launch context won't work since in line 106 that is simply a function signature, what we could do is pass it into "Grid" as an optional parameter. What do you guys think?

File: benchmark/ltimes.cpp
Comment (artv3) [inline benchmark/ltimes.cpp:476]:
```suggestion
    RAJA::launch<pol_launch>(RAJA::ExecPlace::HOST, RAJA::LaunchParams(), [=](RAJA::LaunchContext ctx){
```

File: examples/dynamic_mat_transpose.cpp
Comment (artv3) [inline examples/dynamic_mat_transpose.cpp:144]: For the next release we should create the RAJA alias' for the gpu policies

**Comment by artv3** [comment]
@MrBurmark @trws @homerdin @rhornung67 @rchen20 , this PR is ready for review.

File: examples/dynamic_mat_transpose.cpp
Comment (trws) [reply examples/dynamic_mat_transpose.cpp:139]: What do you mean here? The generic ones or something else?

Comment (artv3) [reply examples/dynamic_mat_transpose.cpp:139]: I figure most codes will have this switch-yard, so we can help codes a bit by doing it for them.

File: docs/sphinx/user_guide/feature/loop_basic.rst
Comment (rchen20) [inline docs/sphinx/user_guide/feature/loop_basic.rst:24]:
```suggestion
.. note:: The ``forall`` ,  ``kernel``, and ``launch`` methods are in the ``RAJA``
          namespace, while ``dynamic_forall`` is in the RAJA namespace for
```

File: docs/sphinx/user_guide/feature/loop_basic.rst
Comment (rchen20) [inline docs/sphinx/user_guide/feature/loop_basic.rst:26]:
```suggestion
          experimental features ``RAJA``.  ``RAJA::expt::dynamic_forall``
          will be moved to the ``RAJA`` namespace in a future RAJA release.
```

File: docs/sphinx/user_guide/feature/loop_basic.rst
Comment (rchen20) [inline docs/sphinx/user_guide/feature/loop_basic.rst:25]:
```suggestion
          experimental features ``RAJA::expt``.  ``RAJA::expt::dynamic_forall``
```

File: docs/sphinx/user_guide/feature/loop_basic.rst
Comment (rchen20) [inline docs/sphinx/user_guide/feature/loop_basic.rst:298]: Delete this?

File: docs/sphinx/user_guide/tutorial/launch_basic.rst
Comment (rchen20) [inline docs/sphinx/user_guide/tutorial/launch_basic.rst:29]: Delete?

File: include/RAJA/pattern/launch/launch_core.hpp
Comment (rchen20) [inline include/RAJA/pattern/launch/launch_core.hpp:479]: Do these `loop` and `tile` policies need to be taken out of `::expt`?

File: docs/sphinx/user_guide/feature/loop_basic.rst
Comment (rchen20) [inline docs/sphinx/user_guide/feature/loop_basic.rst:299]:
```suggestion
```

File: include/RAJA/pattern/launch/launch_core.hpp
Comment (artv3) [reply include/RAJA/pattern/launch/launch_core.hpp:479]: No, these were an initial prototype for perfectly nested loops

File: docs/sphinx/user_guide/tutorial/launch_basic.rst
Comment (rchen20) [inline docs/sphinx/user_guide/tutorial/launch_basic.rst:30]:
```suggestion
hierarchical loop-based parallelism. Kernel execution details 
with RAJA Launch occur inside the lambda expression
```

**Review by rchen20 — APPROVED** [review]
There are a couple other experimental documentation sentences which should be removed or edited, and there are a couple `loop` and `tile` policies still in the `::expt` namespace. Otherwise, everything else looks good.

**Comment by artv3** [comment]
> There are a couple other experimental documentation sentences which should be removed or edited, and there are a couple `loop` and `tile` policies still in the `::expt` namespace. Otherwise, everything else looks good.

Thanks, yes the versions for perfectly nested loops will still be expt until we sort out the interface.

File: docs/sphinx/user_guide/tutorial/launch_basic.rst
Comment (rhornung67) [inline docs/sphinx/user_guide/tutorial/launch_basic.rst:34]:
```suggestion
                     RAJA::Threads(Nthreads,Nthreads)),
```

File: docs/sphinx/user_guide/tutorial/launch_basic.rst
Comment (rhornung67) [inline docs/sphinx/user_guide/tutorial/launch_basic.rst:65]:
```suggestion
programming models, this defines an hierarchical structure in which outer loops 
```

File: docs/sphinx/user_guide/tutorial/launch_exec_pols.rst
Comment (rhornung67) [inline docs/sphinx/user_guide/tutorial/launch_exec_pols.rst:120]:
```suggestion
With the OpenMP version above, ``RAJA::launch`` method is templated on
```

File: docs/sphinx/user_guide/tutorial/launch_exec_pols.rst
Comment (rhornung67) [inline docs/sphinx/user_guide/tutorial/launch_exec_pols.rst:151]:
```suggestion
``RAJA::loop`` methods support execution policies, which enable mapping 
```

File: docs/sphinx/user_guide/tutorial/matrix_transpose.rst
Comment (rhornung67) [inline docs/sphinx/user_guide/tutorial/matrix_transpose.rst:110]:
```suggestion
A CUDA ``RAJA::launch`` variant for the GPU is similar with CUDA
```

File: docs/sphinx/user_guide/tutorial/matrix_transpose_tiled.rst
Comment (rhornung67) [inline docs/sphinx/user_guide/tutorial/matrix_transpose_tiled.rst:136]:
```suggestion
CUDA policies in the ``RAJA::loop`` methods. The complete
```

**Comment by rhornung67** [comment]
@artv3 I activated this branch on RTD so we can review the generated sphinx docs: https://readthedocs.org/projects/raja/versions/

**Comment by artv3** [comment]
The mac build timed out...

**Comment by artv3** [comment]
I reworked the headers in the atomics files, but it seems to be breaking some of the builds. Looking at ways to reproduce it locally.

**Comment by artv3** [comment]
We got a time out for the Docker nvcc11 build

**Comment by artv3** [comment]
yes!!! all green!

**Review by rhornung67 — APPROVED** [review]

File: include/RAJA/pattern/launch/launch_core.hpp
Comment (MrBurmark) [inline include/RAJA/pattern/launch/launch_core.hpp:204]:
```suggestion
#if defined(RAJA_DEVICE_CODE) && defined(RAJA_ENABLE_SYCL)
```

**Comment by rhornung67** [comment]
@artv3 the plan is to merge this before https://github.com/LLNL/RAJA/pull/1004. Do others want to take a final review pass before merging?

File: include/RAJA/pattern/launch/launch_core.hpp
Comment (mdavis36) [inline include/RAJA/pattern/launch/launch_core.hpp:45]: What is NUM_PLACES used for?

Comment (rhornung67) [reply include/RAJA/pattern/launch/launch_core.hpp:45]: I think it's a "count" data member. If kept at the end of the enum, it will always be equal to the number of exec places. Probably used as a loop bound or check in the code.

**Comment by artv3** [comment]
> @artv3 the plan is to merge this before #1004. Do others want to take a final review pass before merging?

Either works, and yes please if folks have a second to take a look that would be great.

File: include/RAJA/pattern/launch/launch_core.hpp
Comment (mdavis36) [reply include/RAJA/pattern/launch/launch_core.hpp:45]: Ahh, fair enough.

**Review by mdavis36 — APPROVED** [review]

**Comment by artv3** [comment]
Did anyone want to take a look?

**Pull request merged**
